dspace-server-webapp/src/main/java/org/dspace/app/rest/repository/CommunitySubcommunityLinkRepository.java

```diff
@@ -46,10 +58,24 @@
             if (community == null) {
                 throw new ResourceNotFoundException("No such community: " + communityId);
             }
-            List<Community> subcommunities = community.getSubcommunities();
-            return converter.toRestPage(utils.getPage(subcommunities, optionalPageable), projection);
-        } catch (SQLException e) {
-            throw new RuntimeException(e);
+            Pageable pageable = utils.getPageable(optionalPageable);
+            List<Community> publicSubcommunities = new LinkedList<Community>();
+            IndexObjectFactoryFactory indexObjectFactory = IndexObjectFactoryFactory.getInstance();
+            IndexableObject scopeObject = indexObjectFactory.getIndexableObjects(context, community).get(0);
+            DiscoverQuery discoverQuery = new DiscoverQuery();
```

Comment:
Same here, please add a comment to describe the query.  This looks to be searching for all Subcommunities which the current user has access to under a given Community.

UPDATE: Is part of this query missing?  I'm not seeing the `location.parent` filter here to ensure we are filtering to subcommunities under one community.  In other words, I'd expect this query to be similar to the one in `CommunityCollectionLinkRepository`

Reply:
@tdonohue you are right... investigating on that we have discovered that a proper test for the `findAllSubCommunities` normal scenario (i.e. all public resources) is missing or better the one that already exists is not related to what we expect, see https://github.com/DSpace/DSpace/blob/880ed67732fcaf7358547185d0bb77befa2703ee/dspace-server-webapp/src/test/java/org/dspace/app/rest/CommunityRestRepositoryIT.java#L733

As you can see here it test a search method NOT the subcommunities path. We are going to add the missing test to demostrate the bug the you have found and proof the resolution but we are worrying about this search method, should we retain it? it is obviously a duplicate of the subcommunities path and actually it runs some duplicated code that need to be cleanup as well.
If we want to retain such search method 
https://github.com/DSpace/Rest7Contract/blob/master/communities.md#subcommunities-1
of course we should add IT also for it about the security scenarios.

Reply:
@abollini : I suspect that the `subCommunities` search was added simply because this `location.parent` filter didn't exist (so it was an early attempt at a subcommunities endpoint).  However, I agree that, assuming we feel the `location.parent` filter is the better direction, we should likely **remove** the `subCommunities` search endpoint and just use `/api/core/communities/[uuid]/subcommunities` as the replacement.

We also should keep in mind that this change also might require minor modifications to the Angular UI, if it is using the `subCommunities` search endpoint.

Reply:
@tdonohue @abollini 
on angular side is only used the community's `subcommunities` link, the `core/communities/search/subCommunities` method is not used

Reply:
From the REST contract I also don't see any reasons why both endpoints exist, but there's little info on either endpoint.
Is it possible that one endpoint only returns direct children and the other endpoint also returns indirect children?

Reply:
we checked, both implementation were returning only direct children. So we have removed the search method

Reply:
the search method has been removed see https://github.com/DSpace/DSpace/pull/2707/commits/9ce5c586269ebdde32e87d1d7591be15734f7dd3#diff-38aef90b55ad47e2a09766c3dedbf89dL202
and the test fixed to use the right endpoint
https://github.com/DSpace/DSpace/pull/2707/commits/9ce5c586269ebdde32e87d1d7591be15734f7dd3#diff-6622e3784efcad38a85da05b7dff958fR922
that has shown the issue caught by @tdonohue  here https://github.com/DSpace/DSpace/pull/2707/commits/9ce5c586269ebdde32e87d1d7591be15734f7dd3#diff-61a40e01c781a6ddea21860116a7679fR68